doc/contrib/copy_on_write.md

```diff
@@ -0,0 +1,85 @@
+# Copy-on-Write
+
+The two basic Elektra datastructures `Key` and `KeySet` implement full copy-on-write (COW) functionality.
+If a key or a keyset gets copied, only a shallow copy with references to the original name and value is created.
+Only if it gets modified the necessary memory is actually allocated.
+As a consequence, duplicated keys or keysets only require a fraction of the memory compared to their source counterparts.
+
+## Working principle
+
+The main principle for the copy-on-write approach is simple: seperate the "real" data from the user-facing identifying entity, in our case `Key` and `KeySet`.
+Developers using Elektra only get references to `Key` and `KeySet` objects.
+Those objects themselves do not hold any data.
+They only contain references to other entities holding the data.
+
+The main entities for copy-on-write are depicted in the following diagram:
+
+```mermaid
+classDiagram
+    Key *-- "0..1" KeyName : keyName
+    Key *-- "0..1" KeyData : keyData
+
+    KeySet *-- "0..1" KeySetData : data
+
+    Key o-- "0..1" KeySet : meta
+    KeySetData o-- "*" Key : array
+
+    Key : - uint16_t refs
+    Key : - keyflag_t flags
+
+    KeyName : - uint16_t refs
+    KeyName : - uint16_t flags
+
+    KeyData : - uint16_t refs
+    KeyData : - uint16_t flags
+
+    KeySet : - uint16_t refs
+    KeySet : - ksflag_t flags
+
+    KeySetData : - uint16_t refs
+    KeySetData : - uint16_t flags
+```
```

Comment:
The Diagram should probably have fields for the actual name, value and key array in the respective structs to show what is stored where.

Reply:
If you look at the rendered diagram, the field names are on the relationships.

Reply:
I saw that. I didn't mean the name of the `KeyName` field in `Key`. I meant the `char * name` field in `KeyName`. It's a bit weird that it is absent, when both `refs` and `flags` are shown.

Reply:
Oooh.... you're right. I added them in `KeyName` and `KeyData`. However, the the key array `KeySetData` is represented by a relationship.